library/core/src/marker.rs

```diff
@@ -482,6 +482,28 @@ impl<T: ?Sized> !Sync for *const T {}
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: ?Sized> !Sync for *mut T {}
 
+#[unstable(feature = "gc", issue = "none")]
+#[cfg_attr(not(test), rustc_diagnostic_item = "FinalizerSafe")]
+/// Types for which it is safe to use inside a finalizer.
+///
+/// A `Gc` will finalize values on a separate thread. This means that a value
+/// used inside a `Gc` must have a thread-safe `Drop` implementation. In most
+/// cases, this is already covered if every type used inside the drop method
+/// implements `Send` + `Sync`.
+///
+/// The `FinalizerSafe` trait means that a type is safe to use inside a `Drop`
```

Comment:
I can't quite parse this and I think it doesn't quite covere everything. I *think* it wants to say something like "This marker trait can be explicitly used to inform the type system that a struct's `drop` method is `Send+Sync`-safe even if the struct as a whole is not. For example, if a struct `S` has a single field `f` whose type is not `Send+Sync` but `S`'s `drop` method does not reference that field, then the struct as a whole can be safely marked `FinalizerSafe`. Note that this trait only affects GC finalisation and has no effect outside the `Gc` type."

Reply:
Yeah that's right. I can change the comment to look more like this. I guess the extra bit though which I'm not sure how to add without maybe confusing things is the following.

It's also valid to call `Gc<T>::new` on a `T: !FinalizerSafe`, provided that `T`'s drop method never references one of it's fields which is `!FinalizerSafe`.

The cool part of this PR is that now we check this statically, whereas before I had to manually check every `Drop` method in WLambda was safe.

Reply:
> It's also valid to call Gc<T>::new on a T: !FinalizerSafe, provided that T's drop method never references one of its fields which is !FinalizerSafe.

That doesn't quite make sense to me, but maybe I've misunderstood. If `T` is not `Send+Sync` and not `FinalizerSafe` surely `Gc::new` will just reject it?

In terms of the comment I might tweak it slightly to:

> This marker trait can be explicitly used to inform the type system that a struct's drop method is Send+Sync-safe even if the struct as a whole is not. For example, if a struct S has a single field f whose type is not Send+Sync but S's drop method does not reference that field, then the struct as a whole can be safely marked `FinalizerSafe`.
>
> Note that this trait only has relevance to GC finalisation and `Gc::new`. In other words, `FinalizerSafe` does not affect a struct when it is not embedded in a `Gc`.

Reply:
> That doesn't quite make sense to me, but maybe I've misunderstood. If T is not Send+Sync and not FinalizerSafe surely Gc::new will just reject it?

It will reject it only if it can't see into its drop method. For example, the following program would be accepted, even though `A: !FinalizerSafe`:

```rust

// A is !Sync and !FinalizerSafe because it contains a `Cell`. 
struct A {
   field1: usize,
   field2: Cell<usize>
}

impl Drop for A {
  fn drop(&mut self) {
    println!("Hello world")
  }
}

fn main() {
  let a = A{...};
  Gc::new(a);
}
```

This is fine because `A::Drop` doesn't dereference anything that is `!FinalizerSafe`. However, if we changed `A`'s drop method to the following:

```rust
impl Drop for A {
  fn drop(&mut self) {
    println!("Hello {}", self.field2.get())
  }
}

```

Now we'd get the following compiler error:

```

error: `a` cannot be safely finalized.
  --> src/main.rs:27:13
   |
18 |         println!("Hello {:?}", self.field2.get())
   |                                -----------------
   |                                |
   |                                caused by the expression in `fn drop(&mut)` here because
   |                                it uses a type which is not safe to use in a finalizer.
...
27 |     Gc::new(a);
   |             ^ has a drop method which cannot be safely finalized.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values whose types implement `Send + Sync` or `FinalizerSafe`.
```

However, if `A` was a type that we couldn't see into at compile time (e.g. it's a trait object, or it hasn't been fully monomorphized), then we would have to reject `A` when it's passed into `Gc::new` because it's `!FinalizerSafe` and we don't know what its drop method does.

Reply:
Oh, we really *do* look inside the `drop` method? Is that recursive? e.g. if in a `drop` method I call `f(&mut self)` will `f` be checked to see if it does anything naughty?

Reply:
>Oh, we really do look inside the drop method? 

Yes that's right!

>Is that recursive?

It's not recursive, no. Right now if we call a function that takes a field that it shouldn't touch as an argument then we get a compiler error. Your example is a good one because it takes a `&mut self`, this should also fail because we don't know what `f` will do with `&mut self`. You've discovered a bug, because we should reject this (as we already do if we had a function `g(&mut Cell<usize>)` and passed in `self.field2`. I shall fix this!

Reply:
Even though it's not recursive though, it's already surprisingly effective (on standard library types and WLambda at least) because we attach the pass after inlining and most `Drop` implementations don't really do anything too complicated other than perhaps a decref / dealloc on particular fields.

Reply:
To clarify though: where we can't recurse through a function call to know that it's safe, we reject the program at compile time. So all we're doing here is easing up a bit and letting more (but obviously not all) correct programs through. It's then up to the user to say "I promise I know what I'm doing" by unsafely implementing `FinalizerSafe` on the type passed to `Gc::new`.

Reply:
> So all we're doing here is easing up a bit and letting more (but obviously not all) correct programs through.

Ah, so we've got a safe underapproximation of correct programs? If so, I'm happy with this PR as-is, and the next PR can improve the underapproximation. But if you want to do that in this one PR, that's also OK with me!